### ISSUE TITLE
Prevent worker processing more than 1 job type at a time

### ISSUE THREAD

**Issue by PaulGrimshaw** [Automattic/kue#1037]
I am just investigating Kue as a job processor. However I am stuck with one concept:

Say I have a worker set to process 2 job types.

Code looks something like this:

queue.process("job1", 1, (job, done) => {
    console.log("processing job 1")
})
queue.process("job2", 1, (job, done) => {
    console.log("processing job 2")
})

The problem with this is that if a "job1" and a "job2" are added to the queue at the same time, the above process will start both in parrallel.

In reality I will have "N" job types, each with configurable parrallel runs. I don't want a separate process per job type, as this will be hard to scale. Each process should be able to handle any job type.

How can I make sure that only 1 batch of job types will be processed at a time by any one worker?

**Comment by PaulGrimshaw** [comment]
I have found a bit of a workaround, but it looses some of the native Kue functionality as it treats all jobs as a single type.

Basically I store all the jobs with a type of "job", and have the process watch this.  Each job has a "type" property in it's data, which can be checked at runtime and the relevant job run.

But would much prefer a solution that worked better with the way Kue is meant to be run.

**Comment by manast** [comment]
@PaulGrimshaw I am not sure I understand your requirement. Do you mean that you want to place different job types in parallel, but that they are executed sequentially? (of course this will change if you have several workers or concurrency levels).

**Comment by PaulGrimshaw** [comment]
Basically i want all jobs in one queue, regardless of type.  I have mulitple workers watching this queue, and each worker should take the next job and run it.

**Comment by manast** [comment]
so in other words, it should work as if you had only one process function where you switch to different process functions depending on the job type.

**Comment by PaulGrimshaw** [comment]
Yes exactly.  This way different job types could be added to the queue and prioritised accordingly, and multiple workers could pick jobs of the queue

**Comment by PaulGrimshaw** [comment]
In agenda, multiple processes watch one queue, and process one job at a time. You can restrict the job types that one process watches, and it will ignore other types.

**Comment by supermafete** [comment]
I'm considering switching to Agenda just because of this. I need to process one job at a time, no matter what type is it.

Is there any way to achieve that?

My case is a little bit different. I have two types of jobs, let's say 'process contents' and 'build relations'. I need to run them one at a time, but they can be runned in any time and in any order. For example 2x process contents, 3x build relations, 1x process contents, etc.

I have decided to create only one generic job type called 'API job'. That means I have to change the .process function even is there is a job running. If I call .process twice, that makes another job added to the active queue, no matter its type nor if it's another job running.
